src/v/net/connection_rate.h

```diff
@@ -29,8 +31,105 @@ using namespace std::chrono_literals;
 namespace net {
 
 struct connection_rate_info {
-    int64_t max_connection_rate;
+    std::optional<int64_t> max_connection_rate;
     std::vector<ss::sstring> overrides;
 };
 
-} // namespace net
\ No newline at end of file
+enum class should_spawn_fiber_on_update { yes, no };
+
+struct connection_rate_counter {
+    explicit connection_rate_counter(int64_t max_rate) noexcept
+      : max_tokens(max_rate)
+      , one_token_time(std::max(1000 / max_tokens, 1l))
+      , rate(max_rate)
+      , last_update_time(ss::lowres_clock::now()) {}
+
+    void broke_semaphore() { rate.broken(); }
+
+    should_spawn_fiber_on_update update_max_rate(int64_t new_max_rate) {
+        auto diff = new_max_rate - max_tokens;
+        if (diff == 0) {
+            return should_spawn_fiber_on_update::no;
+        }
+
+        max_tokens = new_max_rate;
+        one_token_time = std::chrono::milliseconds(
+          std::max(1000 / max_tokens, 1l));
+
+        if (diff > 0) {
+            auto need_add = max_tokens - rate.current();
+            rate.signal(need_add);
+        } else {
+            auto need_move = static_cast<int64_t>(rate.current()) - max_tokens;
+            if (need_move > 0) {
+                rate.consume(need_move);
+            }
+            return should_spawn_fiber_on_update::yes;
+        }
+
+        return should_spawn_fiber_on_update::no;
+    }
+
+    void
+    update_rate(int64_t spawned_tokens, ss::lowres_clock::time_point time) {
+        rate.signal(spawned_tokens);
+        last_update_time = time;
+    }
+
+    int64_t avaiable_new_connections() {
+        return static_cast<int64_t>(rate.current());
+    }
+
+    int64_t max_tokens;
+    std::chrono::milliseconds one_token_time;
+    ss::semaphore rate;
```

Comment:
nit: `rate` is a little bit of a consuming name, because the number in this semaphore isn't the rate ("tokens per second") but is an absolute number of tokens.  Could call it `tokens` or `bucket` for the token bucket idiom?